core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -423,21 +510,30 @@ private[spark] class ExecutorAllocationManager(
    */
   private def removeExecutors(executors: Seq[String]): Seq[String] = synchronized {
     val executorIdsToBeRemoved = new ArrayBuffer[String]
-
     logDebug(s"Request to remove executorIds: ${executors.mkString(", ")}")
-    val numExistingExecutors = executorMonitor.executorCount - executorMonitor.pendingRemovalCount
-
-    var newExecutorTotal = numExistingExecutors
+    val numExecutorsTotalPerRpId = mutable.Map[Int, Int]()
     executors.foreach { executorIdToBeRemoved =>
-      if (newExecutorTotal - 1 < minNumExecutors) {
-        logDebug(s"Not removing idle executor $executorIdToBeRemoved because there are only " +
-          s"$newExecutorTotal executor(s) left (minimum number of executor limit $minNumExecutors)")
-      } else if (newExecutorTotal - 1 < numExecutorsTarget) {
-        logDebug(s"Not removing idle executor $executorIdToBeRemoved because there are only " +
-          s"$newExecutorTotal executor(s) left (number of executor target $numExecutorsTarget)")
+      val rpId = getResourceProfileIdOfExecutor(executorIdToBeRemoved)
+      if (rpId == UNKNOWN_RESOURCE_PROFILE_ID) {
+        logWarning(s"Not removing executor $executorIdsToBeRemoved because couldn't find " +
+          "ResourceProfile for it!")
```

Comment:
so if events come out of order in the ExecutorMonitor we aren't sure what resource profile they are for, or if the executor monitor just doesn't know about the executor, in both of those cases its returns the UNKNOWN. When I wrote this code I was not expecting it to normally happen but actually now that I look more  I think there is a possible race where the remove event could come at the same time as we calculate the timed out executors. The remove event would cause executor monitor to remove it from the list and then you would get UNKNOWN back.  The calculation of timed out executors which calls into this removeExecutors takes a snapshot of the executors so it could return one that has later been removed.  I think I'll modify this to pass the ResourceProfile in with the list of executors to remove, then we don't have to worry about that race.  thanks for bringing this up

So currently we are not cleaning up if all resource profiles have been gc'ed. The executors will idle timeout - down to the minimum set. note I think you saw that discussion but we also want to make the initial, minimum, max configs settable per resource profile - but was going to do that in followon as well.
And many of the datastructures do not remove resource profiles because its not completely obvious when they are done being used. My  thought was the number of ResourceProfiles is expected to be small, I tried to keep the datastructures memory usage small, and was thinking we can make this better in follow on jiras.

I was thinking the main initial use case was the ETL -> ML type case so was thinking you would only have a couple ResourceProfiles. 

If you think its needed now, I can definitely start looking more into this. Just let me know your thoughts.
I filed https://issues.apache.org/jira/browse/SPARK-30749 and https://issues.apache.org/jira/browse/SPARK-30750 to specifically track doing that.

Reply:
Note that on the datastructures we don't clean up, the idea was there would be pretty small, so just simple Int to Int or Int to empty set. If you see somewhere that we leak more then that please point it out.

Reply:
Essentially where ever we are introducing resource profile, a Set[E] becomes a Map[Int, Set[E]] and an Int becomes a Map[Int, Int]. Under normal circumstances, I do not expect these to be bad.
But given that we dont have named resource profiles, I am not sure how it interacts with a loop. For example, in an ML loop :
```
val inputRdd = prepare()
while (condition) {
  val computeRdd = needGpuResources(inputRdd.map().foo.bar)
  ...
}
```

Here, if I understood, for each iteration we will create a new resource profile. Depending on number of iterations, we could end up with an increasing memory usage (degenerate case - if this is used in streaming).
Note: this is not a regression - for default profile, there is no usage increase (other than a negligible increase).

Reply:
Btw, agree about out of order - thx for the details.

Reply:
> Essentially where ever we are introducing resource profile, a Set[E] becomes a Map[Int, Set[E]] and an Int becomes a Map[Int, Int]. Under normal circumstances, I do not expect these to be bad.
> But given that we dont have named resource profiles, I am not sure how it interacts with a loop. For example, in an ML loop :
> 
> ```
> val inputRdd = prepare()
> while (condition) {
>   val computeRdd = needGpuResources(inputRdd.map().foo.bar)
>   ...
> }
> ```
> 
> Here, if I understood, for each iteration we will create a new resource profile. Depending on number of iterations, we could end up with an increasing memory usage (degenerate case - if this is used in streaming).
> Note: this is not a regression - for default profile, there is no usage increase (other than a negligible increase).


So you would not create a new resource profile on just looping unless you are actually building it because you need different requirements on each iteration. You build it once and just re-use that profile.

```
val resourceProfileBuilder = new ResourceProfileBuilder()
val ereq = new ExecutorResourceRequests()
val treq = new TaskResourceRequests()
ereq.cores(2).memory("6g").memoryOverhead("2g")..resource("gpu", 1, "/home/tgraves/getGpus")
treq.cpus(2).resource("gpu", 1)

resourceProfileBuilder.require(ereq)
resourceProfileBuilder.require(treq)
val resourceProfile = resourceProfileBuilder.build()

val inputRdd = prepare()
while (condition) {
   val computeRDD = inputRdd.map().withResources(resourceProfile).foo.bar)
   ....
}
```

In the example above, it creates a single ResourceProfile - it only has 1 id with it.  The id is numeric at this point but its really no different then a name from a uniqueness point of view.

Or maybe your intention was that needGpuResources() function would add it and if that function was creating a profile each time then yes it would but I don't see how that is different whether you have a name or not. If the user does something inefficient all we can do is try to educate them.  I guess the only difference is if they try to create the same profile with the same name - then you probably error but either way you went through all the work of constructing it.
The user should just do it outside the loop like my example.
but does come back to the point about memory, where it could start to add up if the user did something like that.

Reply:
This is the saner way to define it - definitely.
I am only worried about folks doing this within needGpuResources (in example above) - it could be coming from different libraries/teams.
With named profiles, this could either be reference to externally defined profile (by admin for example) or gets deduped into same if equality match (which should be the case here ...).